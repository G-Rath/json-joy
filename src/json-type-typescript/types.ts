/** A module declaration, e.g. "namespace Foo {". */
export interface TsModuleDeclaration {
  node: 'ModuleDeclaration';
  name: string;
  statements: TsDeclaration[];
  comment?: string;
}

/** An interface declaration, e.g. "interface Bar {". */
export interface TsInterfaceDeclaration {
  node: 'InterfaceDeclaration';
  name: string;
  members: TsPropertySignature[];
  comment?: string;
}

/** A property of an interface type. */
export interface TsPropertySignature {
  node: 'PropertySignature';
  name: string;
  optional?: boolean;
  type: TsType;
}

/** A type alias declaration, e.g. "type Baz = ...". */
export interface TsTypeAliasDeclaration {
  node: 'TypeAliasDeclaration';
  name: string;
  type: TsType;
}

/** All possible declarations that can be statements of a module. */
export type TsDeclaration = TsModuleDeclaration | TsInterfaceDeclaration | TsTypeAliasDeclaration;

/** An "Array<*>" type. */
export interface TsArrayType {
  node: 'ArrayType';
  elementType: TsType;
}

/** "string" */
export interface TsStringKeyword {
  node: 'StringKeyword';
}

/** "number" */
export interface TsNumberKeyword {
  node: 'NumberKeyword';
}

/** "boolean" */
export interface TsBooleanKeyword {
  node: 'BooleanKeyword';
}

/** "null" */
export interface TsNullKeyword {
  node: 'NullKeyword';
}

/** "any" */
export interface TsAnyKeyword {
  node: 'AnyKeyword';
}

/** "unknown" */
export interface TsUnknownKeyword {
  node: 'UnknownKeyword';
}

/** Inline interface type. */
export interface TsTypeLiteral {
  node: 'TypeLiteral';
  members: TsPropertySignature[];
}

/** Exact string as type. */
export interface TsStringLiteral {
  node: 'StringLiteral';
  text: string;
}

/** Exact number as type. */
export interface TsNumericLiteral {
  node: 'NumericLiteral';
  text: string;
}

/** "true" */
export interface TsTrueKeyword {
  node: 'TrueKeyword';
}

/** "false" */
export interface TsFalseKeyword {
  node: 'FalseKeyword';
}

/** List of types separated by "|" pipe. */
export interface TsUnionType {
  node: 'UnionType';
  types: TsType[];
}

/** A reference to a type alias, e.g. "foo: Reference". */
export interface TsTypeReference {
  node: 'TypeReference';
  typeName: string;
}

/** All type annotations. */
export type TsType =
  | TsArrayType
  | TsStringKeyword
  | TsNumberKeyword
  | TsBooleanKeyword
  | TsNullKeyword
  | TsAnyKeyword
  | TsTypeLiteral
  | TsStringLiteral
  | TsNumericLiteral
  | TsTrueKeyword
  | TsFalseKeyword
  | TsUnknownKeyword
  | TsUnionType
  | TsTypeReference;

/** Any possible TypeScript AST node. */
export type TsNode = TsDeclaration | TsPropertySignature | TsType;
