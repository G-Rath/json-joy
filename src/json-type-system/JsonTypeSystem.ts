import type {TType} from '../json-type/types';
import {BooleanValidator, createBoolValidator, ObjectValidator, createObjValidator} from '../json-type-validator';
import {dynamicFunction} from '../util/codegen/dynamicFunction';
import {JsonEncoderFn, JsonSerializerCodegen} from '../json-type-serializer';

export type Types = {[ref: string]: TType};

const literalTypes: string[] = ['bin', 'bool', 'nil', 'num', 'str'];

export interface JsonTypeSystemOptions<T extends Types> {
  types: T;
}

export class JsonTypeSystem<T extends Types> {
  public constructor(protected readonly options: JsonTypeSystemOptions<T>) {}

  public readonly ref = (ref: string): TType => {
    const type = this.options.types[ref];
    if (!type) throw new Error(`Type [ref = ${ref}] not found.`);
    return type;
  };

  protected readonly fastValidatorCache: {[ref: string]: BooleanValidator} = {};
  protected readonly fastValidatorUsage: {[ref: string]: number} = {};

  public getFastValidator(ref: string) {
    const type = this.ref(ref);
    const cachedFastValidator = this.fastValidatorCache[ref];
    if (cachedFastValidator) {
      this.fastValidatorUsage[ref] = (this.fastValidatorUsage[ref] || 0) + 1;
      return cachedFastValidator;
    }
    const [validator, setValidator] = dynamicFunction<BooleanValidator>(() => {
      throw new Error(`Type [ref = ${ref}] not implemented.`);
    });
    this.fastValidatorCache[ref] = validator;
    const realValidator = createBoolValidator(type, {
      customValidators: [],
      ref: (id: string) => this.getFastValidator(id),
      skipObjectExtraFieldsCheck: true,
      unsafeMode: true,
    });
    if (this.fastValidatorUsage[ref]) setValidator(realValidator);
    this.fastValidatorCache[ref] = realValidator;
    return realValidator;
  }

  protected readonly fullValidatorCache: {[ref: string]: ObjectValidator} = {};

  public getFullValidator(ref: string) {
    const type = this.ref(ref);
    const cachedFastValidator = this.fastValidatorCache[ref];
    if (cachedFastValidator) return cachedFastValidator;
    const [validator, setValidator] = dynamicFunction<ObjectValidator>(() => {
      throw new Error(`Type [ref = ${ref}] not implemented.`);
    });
    this.fullValidatorCache[ref] = validator;
    setValidator(createObjValidator(type, {
      customValidators: [],
      ref: (id: string) => this.getFastValidator(id),
    }));
    return validator;
  }

  protected readonly jsonSerializerCache: {[ref: string]: JsonEncoderFn} = {};
  protected readonly jsonSerializerUsage: {[ref: string]: number} = {};

  public getJsonSerializer(ref: string): JsonEncoderFn {
    const type = this.ref(ref);
    const jsonSerializerCached = this.jsonSerializerCache[ref];
    if (jsonSerializerCached) {
      this.jsonSerializerUsage[ref] = (this.jsonSerializerUsage[ref] || 0) + 1;
      return jsonSerializerCached;
    }
    const codegen = new JsonSerializerCodegen({
      type,
      ref: (id: string) => {
        const type = this.ref(id);
        if (literalTypes.indexOf(type.__t) !== -1) return type;
        return this.getJsonSerializer(id);
      },
    });
    const [serializer, setSerializer] = dynamicFunction<JsonEncoderFn>(() => {
      throw new Error(`Type [ref = ${ref}] JSON serializer not implemented.`);
    });
    this.jsonSerializerCache[ref] = serializer;
    const realSerializer = codegen.run().compile();
    if (this.jsonSerializerUsage[ref]) setSerializer(realSerializer);
    this.jsonSerializerCache[ref] = realSerializer;
    return realSerializer;
  }
}
