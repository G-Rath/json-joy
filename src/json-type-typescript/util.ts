import type {TsNode} from './types';

export const TAB = '  ';

export const keywords: Record<string, 1 | undefined> = {
  break: 1,
  case: 1,
  catch: 1,
  class: 1,
  const: 1,
  continue: 1,
  debugger: 1,
  default: 1,
  delete: 1,
  do: 1,
  else: 1,
  enum: 1,
  export: 1,
  extends: 1,
  false: 1,
  finally: 1,
  for: 1,
  function: 1,
  if: 1,
  import: 1,
  in: 1,
  instanceof: 1,
  new: 1,
  null: 1,
  return: 1,
  super: 1,
  switch: 1,
  this: 1,
  throw: 1,
  true: 1,
  try: 1,
  typeof: 1,
  var: 1,
  void: 1,
  while: 1,
  with: 1,
  as: 1,
  implements: 1,
  interface: 1,
  let: 1,
  package: 1,
  private: 1,
  protected: 1,
  public: 1,
  static: 1,
  yield: 1,
  any: 1,
  boolean: 1,
  constructor: 1 as 1,
  declare: 1,
  get: 1,
  module: 1,
  require: 1,
  number: 1,
  set: 1,
  string: 1,
  symbol: 1,
  type: 1,
  from: 1,
  of: 1,
  unknown: 1,
};

export const normalizeKey = (prop: string): string =>
  /^[a-z_][a-z_0-9]*$/i.test(prop) && !keywords[prop] ? prop : JSON.stringify(prop);

export const isSimpleType = ({node}: TsNode): boolean =>
  node === 'NumberKeyword' || node === 'StringKeyword' || node === 'BooleanKeyword';
