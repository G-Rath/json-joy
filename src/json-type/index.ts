import {JsonTypeBuilder} from './JsonTypeBuilder';

export * from './types';
export const t = new JsonTypeBuilder();

