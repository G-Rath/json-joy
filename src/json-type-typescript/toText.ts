import { wordWrap } from '../util/wordWrap';
import {TsNode} from './types';
import {TAB, isSimpleType, normalizeKey} from './util';

const formatComment = (comment: string | undefined, __: string): string => {
  if (!comment) return '';
  const lines = wordWrap(comment, {width: 80 - 3 - __.length});
  return __ + '/**\n' + __ + ' * ' + lines.join('\n' + __ + ' * ') + '\n' + __ + ' */\n';
};

export const toText = (node: TsNode | TsNode[], __: string = ''): string => {
  if (Array.isArray(node)) return node.map((s) => toText(s, __)).join('\n');

  const ____ = __ + TAB;

  switch (node.node) {
    case 'ModuleDeclaration': {
      let out: string = '';
      out += `${__}export namespace ${node.name} {\n`;
      out += toText(node.statements, ____);
      out += `${__}}\n`;
      return out;
    }
    case 'InterfaceDeclaration': {
      const {name, members, comment} = node;
      let out: string = '';
      out += formatComment(comment, __);
      out += `${__}export interface ${name} {\n`;
      out += toText(members, ____);
      out += `\n${__}}\n`;
      return out;
    }
    case 'PropertySignature': {
      const name = normalizeKey(node.name);
      return `${__}${name}${node.optional ? '?' : ''}: ${toText(node.type, __)};`;
    }
    case 'IndexSignature': {
      return `${__}[key: string]: ${toText(node.type, __)};`;
    }
    case 'TypeAliasDeclaration': {
      let out: string = '';
      out += formatComment(node.comment, __);
      out += `${__}export type ${node.name} = ${toText(node.type)};\n`;
      return out;
    }
    case 'ArrayType': {
      const simple = isSimpleType(node.elementType);
      const inner = toText(node.elementType, ____);
      return simple ? `${inner}[]` : `Array<${inner}>`;
    }
    case 'StringKeyword': {
      return 'string';
    }
    case 'NumberKeyword': {
      return 'number';
    }
    case 'BooleanKeyword': {
      return 'boolean';
    }
    case 'NullKeyword': {
      return 'null';
    }
    case 'AnyKeyword': {
      return 'any';
    }
    case 'UnknownKeyword': {
      return 'unknown';
    }
    case 'TypeLiteral': {
      return `{\n${toText(node.members, ____)}\n${__}}`;
    }
    case 'StringLiteral': {
      return JSON.stringify(node.text);
    }
    case 'NumericLiteral': {
      return node.text;
    }
    case 'TrueKeyword': {
      return 'true';
    }
    case 'FalseKeyword': {
      return 'false';
    }
    case 'UnionType': {
      return node.types.map((t) => toText(t, ____)).join(' | ');
    }
    case 'TypeReference': {
      return node.typeName;
    }
  }

  return '';
};
