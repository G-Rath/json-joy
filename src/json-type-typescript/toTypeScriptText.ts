import type {Display, TAnyType} from '../json-type';
import wrap from 'word-wrap';
import {normalizeFiledKey} from './util';

const TAB = '  ';

const formatComment = (node: Display, __: string): string => {
  if (node.title || node.description) {
    if (!node.description) {
      return `${__}/** ${node.title} */\n`;
    }
    let res = '';
    res += `${__}/**\n`;
    if (node.title) res += `${__} * # ${node.title}\n`;
    if (node.title && node.description) res += `${__} *\n`;
    if (node.description) {
      const txt = wrap(node.description);
      const lines = txt.split('\n');
      for (const line of lines) {
        res += `${__} * ${line.trim()}\n`;
      }
    }
    res += `${__} */\n`;
    return res;
  }
  return '';
};

export const toTypeScriptText = (type: TAnyType, tab: string): string => {
  if ((type as any).const) {
    return JSON.stringify((type as any).const);
  }
  switch (type.__t) {
    case 'str': {
      return 'string';
    }
    case 'num': {
      return 'number';
    }
    case 'bool': {
      return 'boolean';
    }
    case 'nil': {
      return 'null';
    }
    case 'arr': {
      return `Array<${toTypeScriptText(type.type as TAnyType, tab + TAB)}>`;
    }
    case 'enum': {
      return `[${type.values.map((t) => toTypeScriptText(t as TAnyType, tab + TAB)).join(', ')}]`;
    }
    case 'obj': {
      const ____ = tab + TAB;
      if (type.fields.length === 0) return 'object';
      let res = '{\n';
      for (let i = 0; i < type.fields.length; i++) {
        const field = type.fields[i];
        const key = normalizeFiledKey(field.key) + (field.isOptional ? '?' : '');
        if (i) if (field.key || field.description) res += `\n`;
        res += formatComment(field, ____);
        res += `${____}${key}: ${toTypeScriptText(field.type as TAnyType, ____)};\n`;
      }
      res += tab + '}';
      return res;
    }
    case 'ref': {
      return type.ref;
    }
    case 'or': {
      return type.types.map((t) => toTypeScriptText(t as TAnyType, tab + TAB)).join(' | ');
    }
    case 'bin': {
      return 'Uint8Array';
    }
    case 'any': {
      return 'any';
    }
    default: {
      return 'any';
    }
  }
};
