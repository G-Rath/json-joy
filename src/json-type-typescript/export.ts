import {TAnyType, TObjectField} from '../json-type';
import {
  TsArrayType,
  TsBooleanKeyword,
  TsDeclaration,
  TsFalseKeyword,
  TsInterfaceDeclaration,
  TsModuleDeclaration,
  TsNullKeyword,
  TsNumberKeyword,
  TsNumericLiteral,
  TsPropertySignature,
  TsStringKeyword,
  TsStringLiteral,
  TsTrueKeyword,
  TsType,
  TsTypeAliasDeclaration,
  TsTypeLiteral,
  TsTypeReference,
  TsUnionType,
  TsUnknownKeyword,
} from './types';

export interface ToTypeScriptAstContext {
  /**
   * Resolves a type reference.
   */
  ref: (id: string) => TAnyType;

  /**
   * This array is mutated an populated with new names types or interfaces.
   */
  statements: TsDeclaration[];
}

interface Identifier {
  name: string;
  namespaces: string[];
}

const parseIdentifier = (id: string): Identifier => {
  const parts = id.split('.');
  return {
    name: parts.pop()!,
    namespaces: parts,
  };
};

const augmentWithComment = (
  type: TAnyType | TObjectField,
  node: TsDeclaration | TsPropertySignature | TsTypeLiteral,
) => {
  if (type.title || type.description) {
    let comment = '';
    if (type.title) comment += '# ' + type.title;
    if (type.title && type.description) comment += '\n\n';
    if (type.description) comment += type.description;
    node.comment = comment;
  }
};

/**
 * Converts JSON Type to TypeScript AST declarations.
 *
 * This function is idempotent, if a declaration already exists
 * in the resulting AST, it will not be generated again.
 */
export const exportDeclaration = (ref: string, ctx: ToTypeScriptAstContext): void => {
  const identifier = parseIdentifier(ref);
  let module: TsModuleDeclaration = {
    node: 'ModuleDeclaration',
    name: '',
    statements: ctx.statements,
  };
  for (const namespace of identifier.namespaces) {
    let innerModule: TsModuleDeclaration = module.statements.find(
      (s) => s.node === 'ModuleDeclaration' && s.name === namespace,
    ) as TsModuleDeclaration;
    if (!innerModule) {
      innerModule = {
        node: 'ModuleDeclaration',
        name: namespace,
        statements: [],
      };
      module.statements.push(innerModule);
    }
    module = innerModule;
  }
  const exists = module.statements.find((s) => s.name === identifier.name);
  if (exists) return;
  const type = ctx.ref(ref);
  switch (type.__t) {
    case 'obj': {
      const node: TsInterfaceDeclaration = {
        node: 'InterfaceDeclaration',
        name: identifier.name,
        members: [],
      };
      module.statements.push(node);
      for (const field of type.fields) {
        const member: TsPropertySignature = {
          node: 'PropertySignature',
          name: field.key,
          type: toTsType(field.type as TAnyType, ctx),
        };
        if (field.isOptional) member.optional = true;
        augmentWithComment(field, member);
        node.members.push(member);
      }
      if (type.unknownFields)
        node.members.push({
          node: 'IndexSignature',
          type: {node: 'UnknownKeyword'},
        });
      augmentWithComment(type, node);
      return;
    }
    default: {
      const node: TsTypeAliasDeclaration = {
        node: 'TypeAliasDeclaration',
        name: identifier.name,
        type: toTsType(type, ctx) as TsType,
      };
      augmentWithComment(type, node);
      module.statements.push(node);
      return;
    }
  }
};

const toTsType = (type: TAnyType, ctx: ToTypeScriptAstContext): TsType => {
  switch (type.__t) {
    case 'str': {
      const node: TsStringLiteral | TsStringKeyword =
        typeof type.const === 'string' ? {node: 'StringLiteral', text: type.const} : {node: 'StringKeyword'};
      return node;
    }
    case 'num': {
      const node: TsNumericLiteral | TsNumberKeyword =
        typeof type.const === 'number'
          ? {node: 'NumericLiteral', text: JSON.stringify(type.const)}
          : {node: 'NumberKeyword'};
      return node;
    }
    case 'bool': {
      const node: TsTrueKeyword | TsFalseKeyword | TsBooleanKeyword =
        typeof type.const === 'boolean'
          ? type.const
            ? {node: 'TrueKeyword'}
            : {node: 'FalseKeyword'}
          : {node: 'BooleanKeyword'};
      return node;
    }
    case 'nil': {
      const node: TsNullKeyword = {node: 'NullKeyword'};
      return node;
    }
    case 'arr': {
      const node: TsArrayType = {
        node: 'ArrayType',
        elementType: toTsType(type.type as TAnyType, ctx),
      };
      return node;
    }
    case 'obj': {
      const node: TsTypeLiteral = {
        node: 'TypeLiteral',
        members: [],
      };
      for (const field of type.fields) {
        const member: TsPropertySignature = {
          node: 'PropertySignature',
          name: field.key,
          type: toTsType(field.type as TAnyType, ctx),
        };
        if (field.isOptional) member.optional = true;
        augmentWithComment(field, member);
        node.members.push(member);
      }
      if (type.unknownFields)
        node.members.push({
          node: 'IndexSignature',
          type: {node: 'UnknownKeyword'},
        });
      augmentWithComment(type, node);
      return node;
    }
    case 'enum': {
      const node: TsUnionType = {
        node: 'UnionType',
        types: type.values.map(toTsLiteral),
      };
      return node;
    }
    case 'or': {
      const node: TsUnionType = {
        node: 'UnionType',
        types: type.types.map((t) => toTsType(t as TAnyType, ctx)),
      };
      return node;
    }
    case 'ref': {
      const node: TsTypeReference = {
        node: 'TypeReference',
        typeName: type.ref,
      };
      exportDeclaration(type.ref, ctx);
      return node;
    }
    default: {
      const node: TsUnknownKeyword = {node: 'UnknownKeyword'};
      return node;
    }
  }
};

const toTsLiteral = (value: string | number | boolean | null | unknown): TsType => {
  switch (typeof value) {
    case 'string': {
      const node: TsStringLiteral = {node: 'StringLiteral', text: value};
      return node;
    }
    case 'number': {
      const node: TsNumericLiteral = {node: 'NumericLiteral', text: value.toString()};
      return node;
    }
    case 'boolean': {
      const node: TsTrueKeyword | TsFalseKeyword = {node: value ? 'TrueKeyword' : 'FalseKeyword'};
      return node;
    }
    default: {
      const node: TsUnknownKeyword = {node: 'UnknownKeyword'};
      return node;
    }
  }
};
