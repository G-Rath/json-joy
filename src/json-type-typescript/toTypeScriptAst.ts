import {TAnyType} from "../json-type";
import {TsDeclaration, TsStringKeyword, TsNumberKeyword, TsBooleanKeyword, TsAnyKeyword, TsTypeAliasDeclaration, TsType, TsInterfaceDeclaration, TsPropertySignature, TsTypeReference} from "./types";

export interface ToTypeScriptAstContext {
  /**
   * Resolves a type reference.
   */
  ref: (id: string) => TAnyType;

  /**
   * This array is mutated an populated with new names types or interfaces.
   */
  statements: TsDeclaration[];
}

export const exportDeclaration = (ref: string, ctx: ToTypeScriptAstContext): void => {
  const type = ctx.ref(ref);
  switch(type.__t) {
    case 'nil':
    case 'bool':
    case 'num':
    case 'str': {
      const node: TsTypeAliasDeclaration = {
        node: 'TypeAliasDeclaration',
        name: ref,
        type: toTypeScriptAst(type, ctx) as TsType,
      };
      ctx.statements.push(node);
      return;
    }
    case 'obj': {
      const node: TsInterfaceDeclaration = {
        node: 'InterfaceDeclaration',
        name: ref,
        members: [],
      };
      ctx.statements.push(node);
      for (const field of type.fields) {
        const member: TsPropertySignature = {
          node: 'PropertySignature',
          name: field.key,
          type: toTypeScriptAst(field.type as TAnyType, ctx),
        };
        node.members.push(member);
      }
      return;
    }
  }
};

/**
 * Converts JSON Type to TypeScript AST.
 * 
 * This function is idempotent, if the output is already available
 * in the resulting AST, it will not generate it again.
 */
export const toTypeScriptAst = (type: TAnyType, ctx: ToTypeScriptAstContext): TsType => {
  switch(type.__t) {
    case 'str': {
      const node: TsStringKeyword = {node: 'StringKeyword'};
      return node;
    }
    case 'num': {
      const node: TsNumberKeyword = {node: 'NumberKeyword'};
      return node;
    }
    case 'bool': {
      const node: TsBooleanKeyword = {node: 'BooleanKeyword'};
      return node;
    }
    case 'ref': {
      const node: TsTypeReference = {
        node: 'TypeReference',
        typeName: type.ref,
      };
      exportDeclaration(type.ref, ctx);
      return node;
    }
    default: {
      const node: TsAnyKeyword = {node: 'AnyKeyword'};
      return node;
    }
  }
};
