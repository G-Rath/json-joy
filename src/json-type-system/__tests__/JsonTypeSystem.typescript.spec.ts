import {JsonTypeSystem} from '../JsonTypeSystem';
import {types, customValidators} from '../../demo/json-type/samples';
import {toText} from '../../json-type-typescript/toText';

test('generates JSON schema for simple string type', () => {
  const system = new JsonTypeSystem({types, customValidators});
  const declarations = system.toTsAst();
  const text = toText(declarations);
  expect(declarations).toMatchSnapshot();
  expect(text).toMatchSnapshot();
});
