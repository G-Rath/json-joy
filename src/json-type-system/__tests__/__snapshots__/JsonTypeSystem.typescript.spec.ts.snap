// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generates JSON schema for simple string type 1`] = `
Array [
  Object {
    "name": "ID",
    "node": "TypeAliasDeclaration",
    "type": Object {
      "node": "StringKeyword",
    },
  },
  Object {
    "comment": "# A user object

Users are entities in the system that represent a human. When user is created, automatically a User entity is assigned to that user.",
    "members": Array [
      Object {
        "name": "type",
        "node": "PropertySignature",
        "optional": true,
        "type": Object {
          "node": "StringLiteral",
          "text": "User",
        },
      },
      Object {
        "name": "op",
        "node": "PropertySignature",
        "optional": true,
        "type": Object {
          "node": "NumericLiteral",
          "text": "-1",
        },
      },
      Object {
        "name": "gid",
        "node": "PropertySignature",
        "type": Object {
          "node": "TypeReference",
          "typeName": "ID",
        },
      },
      Object {
        "name": "id",
        "node": "PropertySignature",
        "type": Object {
          "node": "NumberKeyword",
        },
      },
      Object {
        "name": "name",
        "node": "PropertySignature",
        "optional": true,
        "type": Object {
          "node": "StringKeyword",
        },
      },
      Object {
        "name": "email",
        "node": "PropertySignature",
        "type": Object {
          "node": "StringKeyword",
        },
      },
      Object {
        "name": "timeCreated",
        "node": "PropertySignature",
        "type": Object {
          "node": "NumberKeyword",
        },
      },
      Object {
        "name": "timeUpdated",
        "node": "PropertySignature",
        "type": Object {
          "node": "NumberKeyword",
        },
      },
      Object {
        "name": "scores",
        "node": "PropertySignature",
        "type": Object {
          "elementType": Object {
            "node": "NumberKeyword",
          },
          "node": "ArrayType",
        },
      },
      Object {
        "name": "isActive",
        "node": "PropertySignature",
        "type": Object {
          "node": "BooleanKeyword",
        },
      },
      Object {
        "name": "null",
        "node": "PropertySignature",
        "type": Object {
          "node": "NullKeyword",
        },
      },
      Object {
        "name": "unknown",
        "node": "PropertySignature",
        "type": Object {
          "node": "UnknownKeyword",
        },
      },
      Object {
        "name": "isUser",
        "node": "PropertySignature",
        "type": Object {
          "node": "TrueKeyword",
        },
      },
      Object {
        "name": "isPost",
        "node": "PropertySignature",
        "type": Object {
          "node": "FalseKeyword",
        },
      },
      Object {
        "name": "tags",
        "node": "PropertySignature",
        "type": Object {
          "elementType": Object {
            "node": "UnionType",
            "types": Array [
              Object {
                "node": "StringKeyword",
              },
              Object {
                "node": "NumberKeyword",
              },
            ],
          },
          "node": "ArrayType",
        },
      },
      Object {
        "name": "meta",
        "node": "PropertySignature",
        "type": Object {
          "members": Array [
            Object {
              "node": "IndexSignature",
              "type": Object {
                "node": "UnknownKeyword",
              },
            },
          ],
          "node": "TypeLiteral",
        },
      },
    ],
    "name": "User",
    "node": "InterfaceDeclaration",
  },
  Object {
    "members": Array [
      Object {
        "name": "user",
        "node": "PropertySignature",
        "type": Object {
          "members": Array [
            Object {
              "name": "id",
              "node": "PropertySignature",
              "optional": true,
              "type": Object {
                "node": "NumberKeyword",
              },
            },
            Object {
              "name": "name",
              "node": "PropertySignature",
              "optional": true,
              "type": Object {
                "node": "StringKeyword",
              },
            },
            Object {
              "name": "email",
              "node": "PropertySignature",
              "type": Object {
                "node": "StringKeyword",
              },
            },
          ],
          "node": "TypeLiteral",
        },
      },
    ],
    "name": "CrateUserRequest",
    "node": "InterfaceDeclaration",
  },
  Object {
    "comment": "# A response to a create user request

The response to a create user request.",
    "members": Array [
      Object {
        "name": "user",
        "node": "PropertySignature",
        "type": Object {
          "node": "TypeReference",
          "typeName": "User",
        },
      },
    ],
    "name": "CreateUserResponse",
    "node": "InterfaceDeclaration",
  },
  Object {
    "comment": "# A response to a create user request

The response to a create user request.",
    "members": Array [
      Object {
        "name": "user",
        "node": "PropertySignature",
        "type": Object {
          "node": "TypeReference",
          "typeName": "User",
        },
      },
      Object {
        "name": "changes",
        "node": "PropertySignature",
        "type": Object {
          "node": "NumberKeyword",
        },
      },
    ],
    "name": "UpdateUserResponse",
    "node": "InterfaceDeclaration",
  },
  Object {
    "name": "pubsub",
    "node": "ModuleDeclaration",
    "statements": Array [
      Object {
        "name": "channel",
        "node": "ModuleDeclaration",
        "statements": Array [
          Object {
            "members": Array [
              Object {
                "name": "id",
                "node": "PropertySignature",
                "type": Object {
                  "node": "StringKeyword",
                },
              },
              Object {
                "name": "payload",
                "node": "PropertySignature",
                "type": Object {
                  "node": "TypeReference",
                  "typeName": "pubsub.channel.PayloadType",
                },
              },
              Object {
                "name": "meta",
                "node": "PropertySignature",
                "type": Object {
                  "members": Array [
                    Object {
                      "name": "description",
                      "node": "PropertySignature",
                      "type": Object {
                        "node": "StringKeyword",
                      },
                    },
                  ],
                  "node": "TypeLiteral",
                },
              },
            ],
            "name": "Channel",
            "node": "InterfaceDeclaration",
          },
          Object {
            "name": "PayloadType",
            "node": "TypeAliasDeclaration",
            "type": Object {
              "node": "UnionType",
              "types": Array [
                Object {
                  "node": "StringLiteral",
                  "text": "json",
                },
                Object {
                  "node": "StringLiteral",
                  "text": "text",
                },
                Object {
                  "node": "StringLiteral",
                  "text": "blob",
                },
              ],
            },
          },
          Object {
            "members": Array [
              Object {
                "name": "channel",
                "node": "PropertySignature",
                "type": Object {
                  "node": "TypeReference",
                  "typeName": "pubsub.channel.Channel",
                },
              },
              Object {
                "name": "project",
                "node": "PropertySignature",
                "type": Object {
                  "node": "TypeReference",
                  "typeName": "util.Project",
                },
              },
            ],
            "name": "CreateChannelResponse",
            "node": "InterfaceDeclaration",
          },
        ],
      },
    ],
  },
  Object {
    "name": "util",
    "node": "ModuleDeclaration",
    "statements": Array [
      Object {
        "members": Array [
          Object {
            "name": "id",
            "node": "PropertySignature",
            "type": Object {
              "node": "StringKeyword",
            },
          },
          Object {
            "name": "name",
            "node": "PropertySignature",
            "type": Object {
              "node": "StringKeyword",
            },
          },
        ],
        "name": "Project",
        "node": "InterfaceDeclaration",
      },
    ],
  },
]
`;

exports[`generates JSON schema for simple string type 2`] = `
"export type ID = string;

/**
 * # A user object
 * 
 * Users are entities in the system that represent a human. When user is 
 * created, automatically a User entity is assigned to that user.
 */
export interface User {
  \\"type\\"?: \\"User\\";
  op?: -1;
  gid: ID;
  id: number;
  name?: string;
  email: string;
  timeCreated: number;
  timeUpdated: number;
  scores: number[];
  isActive: boolean;
  \\"null\\": null;
  \\"unknown\\": unknown;
  isUser: true;
  isPost: false;
  tags: Array<string | number>;
  meta: {
    [key: string]: unknown;
  };
}

export interface CrateUserRequest {
  user: {
    id?: number;
    name?: string;
    email: string;
  };
}

/**
 * # A response to a create user request
 * 
 * The response to a create user request.
 */
export interface CreateUserResponse {
  user: User;
}

/**
 * # A response to a create user request
 * 
 * The response to a create user request.
 */
export interface UpdateUserResponse {
  user: User;
  changes: number;
}

export namespace pubsub {
  export namespace channel {
    export interface Channel {
      id: string;
      payload: pubsub.channel.PayloadType;
      meta: {
        description: string;
      };
    }

    export type PayloadType = \\"json\\" | \\"text\\" | \\"blob\\";

    export interface CreateChannelResponse {
      channel: pubsub.channel.Channel;
      project: util.Project;
    }
  }
}

export namespace util {
  export interface Project {
    id: string;
    name: string;
  }
}
"
`;
